chain/block_filterer.go

```diff
@@ -100,21 +101,32 @@ func NewBlockFilterer(params *chaincfg.Params,
 	}
 }
 
+// checkFilterTx will check if the transaction is relevant and recursively loop through the
+// inputs to double check transactions that are in the block but were already processed.
+// This is necessary if the block is not sorted in topological order.
+func (bf *BlockFilterer) checkFilterTx(tx *wire.MsgTx, inputs map[chainhash.Hash]*wire.MsgTx) {
+	if bf.FilterTx(tx) {
+		bf.RelevantTxns = append(bf.RelevantTxns, tx)
+		if dependentTx, ok := inputs[tx.TxHash()]; ok {
+			bf.checkFilterTx(dependentTx, inputs)
+		}
+	}
+}
+
 // FilterBlock parses all txns in the provided block, searching for any that
 // contain addresses of interest in either the external or internal reverse
 // filters. This method return true iff the block contains a non-zero number of
 // addresses of interest, or a transaction in the block spends from outpoints
 // controlled by the wallet.
 func (bf *BlockFilterer) FilterBlock(block *wire.MsgBlock) bool {
-	var hasRelevantTxns bool
+	inputs := make(map[chainhash.Hash]*wire.MsgTx)
 	for _, tx := range block.Transactions {
-		if bf.FilterTx(tx) {
-			bf.RelevantTxns = append(bf.RelevantTxns, tx)
-			hasRelevantTxns = true
+		for _, in := range tx.TxIn {
+			inputs[in.PreviousOutPoint.Hash] = tx
```

Comment:
what if two txs spend from the same input tx?

Reply:
Yeah I am not sure that will work, we should definitely write a test for this case.

Reply:
Maybe it should map to a list of txs?

Reply:
Yeah I think a slice of tx's would be needed. However, we will know if we add the required test for this case. I have a feeling it will be useful to keep around if we ever refactor this code. =)